src/enzyme_ad/jax/Passes/MarkFunctionMemoryEffectsPass.cpp

```diff
@@ -0,0 +1,70 @@
+#include "src/enzyme_ad/jax/Passes/Passes.h"
+
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace enzyme {
+#define GEN_PASS_DEF_MARKFUNCTIONMEMORYEFFECTSPASS
+#include "src/enzyme_ad/jax/Passes/Passes.h.inc"
+} // namespace enzyme
+} // namespace mlir
+
+using namespace mlir;
+using namespace mlir::enzyme;
+
+namespace {
+struct MarkFunctionMemoryEffectsPass
+    : public enzyme::impl::MarkFunctionMemoryEffectsPassBase<
+          MarkFunctionMemoryEffectsPass> {
+
+  void runOnOperation() override {
+    ModuleOp module = getOperation();
+    auto *ctx = module->getContext();
+    OpBuilder builder(ctx);
+
+    module.walk([&](FunctionOpInterface funcOp) {
+      bool hasRead = false;
+      bool hasWrite = false;
+      bool hasAllocate = false;
+      bool hasFree = false;
+
+      funcOp.walk([&](Operation *op) {
+        if (auto memOp = dyn_cast<MemoryEffectOpInterface>(op)) {
```

Comment:
We should be conservative and assume all possible effects if an operation doesn't implement this interface. We can't reason about it.

Reply:
Also consider the `HasRecursiveSideEffects` trait.

Reply:
> We should be conservative and assume all possible effects if an operation doesn't implement this interface. We can't reason about it.

Add a flag for this (mostly for testing that we can get the minimal set correctly) with the default being we assume all possible effects.

> Also consider the HasRecursiveSideEffects trait.

I will check it out